### PR TITLE
Fix Product changes cannot be saved outside Products tab

### LINKED ISSUE #2175: Products: Product details can be edited but not saved outside Products tab
**Describe the bug**

When I tap on a product outside the Products tab (e.g. in order details or Top Performers) I get an editable product. However, when I make changes there is no Update button and when I close the product details my changes are lost.

**To Reproduce**

1. Open the My store tab.
2. In the Top Performers section, select a product.
3. In the product details, edit the product (e.g. change the product name).
4. Notice that there is no Update button. Tap the "X" to close the product details.
5. Select the product again and notice that your changes weren't saved, and there was no "discard changes" warning.
6. Open the Products tab.
7. Select the same product.
8. In the product details, edit the product again and notice that an Update button appears.
9. Tap the back button and notice a "discard changes" warning appears.

**Screenshots**

Products tab (Expected)|My store tab (bug)
-|-
IMG_0367|IMG_0366

Screencast: https://cloudup.com/c68aqus-4zH

**Expected behavior**

I expect to see an Update button after making edits to the product details, so I can save my changes.

I also expect to see a "discard changes" dialog if I try to close the edited product details without saving my changes.

**Mobile Environment**
Please include:
- Device: iPad Mini
- iOS version: iPadOS 13.4.1
- WooCommerce iOS version: 4.1.0.0 TestFlight beta

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #2175 

## Cause of the issue

When the product form (edit product main screen) is shown from tapping a product on the Order details screen or Top Performers UI, we only have the product ID and thus we currently have a container `ProductLoaderViewController` that loads the `Product` model and then displays the product details or product form as a child view controller. However, since we have custom logic for the navigation item in `ProductFormViewController` (like `shouldPopOnBackButton` and the `rightBarButtonItems` where the Update CTA is only shown when there are outstanding changes), the navigation item events/changes from `ProductFormViewController` aren't synced with the container view controller `ProductLoaderViewController` and thus we can't see the Update CTA and discard changes action sheet.

## Proposed fix

I've considered other ways to fix this issue and picked the last one for this PR:
- Make `ProductFormViewController`'s `rightBarButtonItems` observable so that `ProductLoaderViewController` can update its navigation item on change
- Update `ProductLoaderViewController` to be pushed to the navigation stack instead of presented modally (this requires design confirmation)
- [Implemented in this PR] In `ProductLoaderViewController`, push `ProductFormViewController` to the navigation stack without animation & back button instead of adding it as a child view controller

## Changes

- In `UIViewController`'s helper `addCloseNavigationBarButton`, added support for custom target and action
- In `ProductFormViewController`, added `onDismiss` closure to its initializer so that how the view controller is dismissed can be customized as we want to dismiss modally outside of Products tab (like when pushed from `ProductLoaderViewController`), and pop from the navigation stack when pushed from the Products tab (or its search results)
  - Updated the `ProductFormViewController` initializer usage in `ProductsViewController` and `ProductSearchUICommand`
- In `ProductLoaderViewController` (a view controller that fetches the product by product ID), updated the case where we are showing the product form from embedding it as a child view controller to pushing it to the navigation stack with a close button instead

## Testing

Note: please test this PR on simulators/devices that aren't iOS 13.4, otherwise "Discard changes" action sheet wouldn't work (https://github.com/woocommerce/woocommerce-ios/issues/2089)

### Edit Product from outside of Products tab

- Open the My store tab
- In the Top Performers section, select a simple product (if it's not available, try from a simple product in an order in the Orders tab) --> the "Update" CTA should not appear
- Make some changes to the product --> the "Update" CTA should appear
- Tap the "Update" CTA --> the product changes should be saved and the "Update" CTA is hidden
- Make some other changes to the product --> the "Update" CTA should reappear
- Tap the "x" in the navigation bar --> there should be a discard changes action sheet
- Tap "Discard changes" action --> the product form should be dismissed modally

### Edit Product from Products tab to test for regression

- Open the Products tab
- Tap on a simple product --> the "Update" CTA should not appear
- Make some changes to the product --> the "Update" CTA should appear
- Tap the "Update" CTA --> the product changes should be saved and the "Update" CTA is hidden
- Make some other changes to the product --> the "Update" CTA should reappear
- Tap the "<" in the navigation bar --> there should be a discard changes action sheet
- Tap "Discard changes" action --> the product form should be popped

## Example screencast

In the screencast, it also tested tapping on a non-simple product and made sure the dismiss CTA works correctly (modally or popped from the navigation stack).

simulator

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by oguzkocer** [comment]
Hi there! `4.2` is being frozen, so this PR is being bumped to `4.3`. If these changes _have to_ be shipped with `4.2`, please target `release/4.2` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

File: WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift [hunk 1]
```diff
@@ -162,23 +176,29 @@ private extension ProductLoaderViewController {
     }
 
     func presentProductDetails(for product: Product, isEditProductsEnabled: Bool) {
-        let viewController: UIViewController
         if product.productType == .simple && isEditProductsEnabled {
-            viewController = ProductFormViewController(product: product, currency: currency)
+            let productForm = ProductFormViewController(product: product, currency: currency) { viewController in
+                viewController.dismiss(animated: true, completion: nil)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift:181]: It seems that this callback is never called.

File: WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift [hunk 2]
```diff
@@ -162,23 +176,29 @@ private extension ProductLoaderViewController {
...
+            }
+            productFormViewController = productForm
+
+            productForm.navigationItem.hidesBackButton = true
+            productForm.addCloseNavigationBarButton(target: self, action: #selector(productFormCloseButtonTapped))
+
+            navigationController?.pushViewController(productForm, animated: false)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift:188]: Unfortunately pushing the view controller here creates a problem if you try to do a swipe back. I tried to use  
```
navigationController?.setViewControllers([productForm], animated: true)
```
but it seems that in this case, the close button doesn't work anymore 🤔

File: WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/ProductLoaderViewController.swift]: the `onDismiss` closure was called when the user taps on the "discard changes" action 🙂 (it's not called when the user navigates/swipes back)

File: RELEASE-NOTES.txt [hunk 3]
```diff
@@ -8,6 +8,7 @@
 - The Orders list is now automatically refreshed when reopening the app. 
 - The Orders list is automatically refreshed if a new order (push notification) comes in.
 - Orders -> Search: The statuses now shows the total number of orders with that status.
+- Products: now the Product details can be edited and saved outside Products tab (e.g. from Order details or Top Performers).
```
[the remark this answers is not in the document]
Comment (jaclync) [reply RELEASE-NOTES.txt]: yup! moved to 4.3 in: https://github.com/woocommerce/woocommerce-ios/pull/2233/commits/eef3d0b67594c6aad521cdfa82ac42f3b978540d

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift [hunk 4]
```diff
@@ -55,8 +55,11 @@ final class ProductFormViewController: UIViewController {
 
     private let currency: String
 
-    init(product: Product, currency: String) {
+    private let onDismiss: (UIViewController) -> Void
+
+    init(product: Product, currency: String, onDismiss: @escaping (UIViewController) -> Void) {
```
[the remark this answers is not in the document]
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift]: I took a look at what currency is passed to the view controller in the codebase, and noticed that if the product form is navigated from an order (e.g. from refund details or order details) we incorrectly use the order's currency which can be different from the site's currency. I created a separate issue to fix this: https://github.com/woocommerce/woocommerce-ios/issues/2240

**Comment by jaclync** [comment]
@pmusolino thanks for catching the issue with the navigation stack approach without animation! I thought more about it and updated the implementation based on a new `PresentationStyle` enum so that the navigation details are within `ProductFormViewController`.  lemme know what you think!

**Pull request merged**